### PR TITLE
Feat/6 mirror ledger function

### PR DESCRIPTION
**Pull request by AlexTheWizardL**
# Pull Request: Implement and Test Ledger Mirroring Functionality

## Description

This pull request introduces the `mirror_ledger` function to the `CashCtrlLedger` class, enabling synchronization of ledger data with a target DataFrame. It also includes comprehensive tests to validate the functionality.

### Changes

1. **Function Implementation:**
    - Added `mirror_ledger` function to the `CashCtrlLedger` class.
    - Standardizes the ledger and target data.
    - Identifies and removes duplicate transactions.
    - Adds missing transactions from the target to the ledger.
    - Optionally deletes entries that exist in the ledger but not in the target.

2. **Tests:**
    - `test_mirror_ledger` function to ensure the accuracy and robustness of the `mirror_ledger` method.
    - Validates the following scenarios:
        - Mirroring with one single and one collective transaction.
        - Mirroring with duplicate transactions and `delete=False`.
        - Mirroring with `delete=False`.
        - Mirroring with `delete=True`.
        - Restoration of the initial ledger state.

@lasuk Please review the changes and provide feedback or approval for merging.

### CODE REVIEW

File: tests/test_ledger.py [hunk 1]
```diff
@@ -232,3 +232,65 @@ def test_delete_non_existent_ledger():
...
+        collective_transaction
+    ]))
+    target['date'] = target['date'].ffill()
+    cashctrl_ledger.mirror_ledger(target=target)
+    mirrored = cashctrl_ledger.ledger().reset_index(drop=True)
+    m = target.merge(mirrored.drop(columns=['id']), how='left', indicator=True)
+    assert len(target) == len(mirrored)
+    assert (m['_merge'] == 'both').all(), (
+            'Mirroring error: Some target ledger were not mirrored'
+        )
```
Comment (lasuk) [inline tests/test_ledger.py:264]: Testing whether individual transaction items match between both sides is not a sufficient test. In theory, there could be a mismatch between transaction item and ids. A better test is to convert transactions to unique string representations and ensure the number of occurrences is identical on both sides.

Maybe it would be helpful to write a utility function that converts a ledger data frame into unique string representations. This function could be used both in the package module and in tests.

File: tests/test_ledger.py [hunk 2]
```diff
@@ -232,3 +232,65 @@ def test_delete_non_existent_ledger():
...
+
+    # Mirror with delete=False
+    before_mirroring = cashctrl_ledger.ledger().reset_index(drop=True)
```
Comment (lasuk) [inline tests/test_ledger.py:267]: I guess this would be sufficient (and spare an API call):
```
before_mirroring = mirrored
```

File: tests/test_ledger.py [hunk 3]
```diff
@@ -232,3 +232,65 @@ def test_delete_non_existent_ledger():
...
+    cashctrl_ledger.mirror_ledger(target=target, delete=False)
+    mirrored = cashctrl_ledger.ledger().reset_index(drop=True)
+    m = before_mirroring.merge(mirrored.drop(columns=['id']), how='left', indicator=True)
+    assert len(before_mirroring) == len(mirrored)
+    assert (m['_merge'] == 'both').all(), (
+            'Mirroring error: Some target ledger were not mirrored'
+        )
+
+    # Mirror with delete=True
+    target = StandaloneLedger.standardize_ledger(pd.concat([individual_transaction, collective_transaction]))
```
Comment (lasuk) [inline tests/test_ledger.py:279]: I would add an alternative individual or collective transaction here, to make sure that mirroring works with new transactions.

File: tests/test_ledger.py [hunk 4]
```diff
@@ -232,3 +232,65 @@ def test_delete_non_existent_ledger():
...
+    target['date'] = target['date'].ffill()
+    cashctrl_ledger.mirror_ledger(target=target)
+    mirrored = cashctrl_ledger.ledger().reset_index(drop=True)
+    m = target.merge(mirrored.drop(columns=['id']), how='left', indicator=True)
+    assert len(target) == len(mirrored)
+    assert (m['_merge'] == 'both').all(), (
+            'Mirroring error: Some target ledger were not mirrored'
+        )
+
```
Comment (lasuk) [inline tests/test_ledger.py:288]: Please add a test where target is empty and delete=True and confirm remote ledger is empty.

**Review by lasuk — COMMENTED** [review]
Thanks for implementing the mirroring function!

I have left a series of questions and comments. Back to you @AlexTheWizardL.

File: cashctrl_ledger/ledger.py [hunk 5]
```diff
@@ -322,6 +324,42 @@ def update_vat_code(
         }
         self._client.post("tax/update.json", data=payload)
 
+    def mirror_ledger(self, target: pd.DataFrame, delete: bool = True):
+        remote = StandaloneLedger.standardize_ledger(self.ledger())
+        remote['date'] = remote['date'].ffill()
+        remote = nest(remote, columns=[col for col in remote.columns if col not in ['id', 'date']], key='txn')
+        remote['txn_str'] = [f'{str(date)},{df_to_consistent_str(txn)}' for date, txn in zip(remote['date'], remote['txn'])]
+        target = StandaloneLedger.standardize_ledger(target)
+        target['date'] = target['date'].ffill()
```
[the remark this answers is not in the document]
Comment (AlexTheWizardL) [reply cashctrl_ledger/ledger.py:333]: We can simplify to this
```python
target = self.sanitize_ledger(ledger=self.standardize_ledger(target))
```

**Comment by lasuk** [comment]
I've added two commits to your PR

- 9762667 Fixes erroneous ways of accessing the first element of a pandas Series.
- 2edb906 Enhances readability of the mirroring function

@AlexTheWizardL: Please shout if you disagree.

File: tests/test_ledger.py [hunk 6]
```diff
@@ -232,3 +259,66 @@ def test_delete_non_existent_ledger():
...
+    new_collective_transaction.at[0, 'id'] = 6;
+    new_collective_transaction.at[1, 'id'] = 6;
+    target = StandaloneLedger.standardize_ledger(pd.concat([
+        individual_transaction,
+        new_individual_transaction,
+        collective_transaction,
+        new_collective_transaction,
+    ]))
+    cashctrl_ledger.mirror_ledger(target=target)
+    mirrored = cashctrl_ledger.ledger()
```
[the remark this answers is not in the document]
Comment (AlexTheWizardL) [reply tests/test_ledger.py:288]: Because I do this in the `txn_to_srt()` function [line 65](https://github.com/macxred/cashctrl_ledger/blob/2edb906e97472bc055004b692c933c22e424c40f/tests/test_ledger.py#L64)

```python
def txn_to_str(df: pd.DataFrame) -> List[str]:
    df = df.drop(columns=['id']).reset_index(drop=True)
```

**Comment by lasuk** [comment]
Thanks for the changes.

I've pushed commit dbce642 with additional cosmetic changes.
@AlexTheWizardL: Please shout if you disagree.
